### PR TITLE
refactor(repl): move rustyline sync channel communication into struct

### PR DESCRIPTION
**Pull request by dsherret**
I extracted this out of the repl import specifier autocompletion PR I'm working on.

Basically, for that PR I'm going to have to also get messages from the LSP for autocompletion and it would get out of hand quickly to not extract this out.

### CODE REVIEW

**Comment by dsherret** [comment]
Just noticed #12900

I can wait until after that PR is merged to merge this one cc @lucacasonato

**Review by bartlomieju — APPROVED** [review]
LGTM, nice refactor, wish there was async import in one of the "readline" crates.

**Pull request merged**
